internal/links/device.go

```diff
@@ -0,0 +1,168 @@
+/**
+# Copyright 2023 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package links
+
+import (
+	"fmt"
+
+	"github.com/NVIDIA/go-nvlib/pkg/nvlib/device"
+	"github.com/NVIDIA/go-nvlib/pkg/nvml"
+)
+
+type P2PLinkType uint
+
+const (
+	P2PLinkUnknown P2PLinkType = iota
+	P2PLinkCrossCPU
+	P2PLinkSameCPU
+	P2PLinkHostBridge
+	P2PLinkMultiSwitch
+	P2PLinkSingleSwitch
+	P2PLinkSameBoard
+	SingleNVLINKLink
+	TwoNVLINKLinks
+	ThreeNVLINKLinks
+	FourNVLINKLinks
+	FiveNVLINKLinks
+	SixNVLINKLinks
+	SevenNVLINKLinks
+	EightNVLINKLinks
+	NineNVLINKLinks
+	TenNVLINKLinks
+	ElevenNVLINKLinks
+	TwelveNVLINKLinks
+	ThirteenNVLINKLinks
+	FourteenNVLINKLinks
+	FifteenNVLINKLinks
+	SixteenNVLINKLinks
+	SeventeenNVLINKLinks
+	EighteenNVLINKLinks
+)
+
+// GetP2PLink gets the peer-to-peer connectivity between two devices.
+func GetP2PLink(dev1 device.Device, dev2 device.Device) (P2PLinkType, error) {
+	level, ret := dev1.GetTopologyCommonAncestor(dev2)
+	if ret != nvml.SUCCESS {
+		return P2PLinkUnknown, fmt.Errorf("failed to get commmon anscestor: %v", ret)
+	}
+
+	switch level {
+	case nvml.TOPOLOGY_INTERNAL:
+		return P2PLinkSameBoard, nil
+	case nvml.TOPOLOGY_SINGLE:
+		return P2PLinkSingleSwitch, nil
+	case nvml.TOPOLOGY_MULTIPLE:
+		return P2PLinkMultiSwitch, nil
+	case nvml.TOPOLOGY_HOSTBRIDGE:
+		return P2PLinkHostBridge, nil
+	case nvml.TOPOLOGY_NODE: // NVML_TOPOLOGY_CPU was renamed NVML_TOPOLOGY_NODE
+		return P2PLinkSameCPU, nil
+	case nvml.TOPOLOGY_SYSTEM:
+		return P2PLinkCrossCPU, nil
+
+	}
+
+	return P2PLinkUnknown, fmt.Errorf("unknown topology level: %v", level)
+}
+
+func GetNVLink(dev1 device.Device, dev2 device.Device) (P2PLinkType, error) {
```

Comment:
nit: this public function is missing a comment

Reply:
Updated.